File: DSA/done/953 - isAlienSorted.py
# https://leetcode.com/problems/verifying-an-alien-dictionary/

from typing import List

class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        hm = {o: i for i, o in enumerate(order)}

        for i in range(1, len(words)):
            w1, w2 = words[i - 1], words[i]
            for j in range(len(w1)):
                if j == len(w2):
                    return False
                if w1[j] != w2[j]:
                    if hm[w1[j]] > hm[w2[j]]:
                        return False
                    break
        
        return True 

File: DSA/done/953 - isAlienSorted.cpp
// https://leetcode.com/problems/verifying-an-alien-dictionary/

#include <string>
#include <vector>

using namespace std;
 
class Solution {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        int orderIndex[26] = {0};
        for (int i = 0; i < order.size(); ++i)
            orderIndex[order[i] - 'a'] = i;
        
        for (int i = 1; i < words.size(); i++) {
            string w1 = words[i - 1], w2 = words[i];
            for (int j = 0; j < w1.size(); j++) {
                if (j == w2.size()) 
                    return false;
                if (w1[j] != w2[j]) {
                    if (orderIndex[w1[j] - 'a'] > orderIndex[w2[j] - 'a'])
                        return false;
                    break;
                }
            }
        }

        return true;
    }
};